hazelcast/src/main/java/com/hazelcast/spi/impl/AbstractCompletableFuture.java

```diff
@@ -130,6 +131,13 @@ public boolean cancel(boolean mayInterruptIfRunning) {
                 return false;
             }
 
+            if (shouldCancel == null) {
+                shouldCancel = shouldCancel(mayInterruptIfRunning);
+                if (!shouldCancel) {
+                    return false;
+                }
+            }
+
```

Comment:
Argh, this wasn't actually meant to be like this. Thx for spotting.  I fixed it - it's called once only but the result should be saved for next iterations should the first one fail to set the result.
```
if (shouldCancel == null) {
                shouldCancel = shouldCancel(mayInterruptIfRunning);
            }
            if (!shouldCancel) {
                return false;
            }
```